### PR TITLE
Bug 1854641 - The downloads should stop when the user closes the private tabs.

### PR DESCRIPTION
**Pull request by iorgamgabriel**
### Pull Request checklist

- [x] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [x] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/firefox-android/blob/main/docs/changelog.md) or does not need one
- [x] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/firefox-android/blob/main/docs/shared/android/accessibility_guide.md) or does not include any user facing features

### After merge
- **Breaking Changes**: If this is a breaking Android Components change, please push a draft PR on [Reference Browser](https://github.com/mozilla-mobile/reference-browser) to address the breaking issues.

### To download an APK when reviewing a PR (after all CI tasks finished running):
1. Click on `Checks` at the top of the PR page.
2. Click on the `firefoxci-taskcluster` group on the left to expand all tasks.
3. Click on the `build-apk-{fenix,focus,klar}-debug` task you're interested in.
4. Click on `View task in Taskcluster` in the new `DETAILS` section.
5. The APK links should be on the right side of the screen, named for each CPU architecture.

### GitHub Automation
https://bugzilla.mozilla.org/show_bug.cgi?id=1854641

### CODE REVIEW

**Comment by iorgamgabriel** [comment]
https://github.com/mozilla-mobile/firefox-android/assets/8118371/03d79b10-ceec-40e3-8dba-55c636f3fe07

File: android-components/components/feature/downloads/src/main/java/mozilla/components/feature/downloads/AbstractFetchDownloadService.kt [hunk 1]
```diff
@@ -267,6 +267,27 @@ abstract class AbstractFetchDownloadService : Service() {
...
+                    it.lastNotificationUpdate = System.currentTimeMillis()
+                    setDownloadJobStatus(it, CANCELLED)
+                    it.job?.cancel()
+
+                    it.job = CoroutineScope(IO).launch {
+                        deleteDownloadingFile(it.state)
+                        it.downloadDeleted = true
+                    }
+                    removeDownloadJob(it)
+                }
```
Comment (Amejia481) [inline android-components/components/feature/downloads/src/main/java/mozilla/components/feature/downloads/AbstractFetchDownloadService.kt:289]: This looks really similar to what we are doing for [ACTION_CANCEL](https://github.com/mozilla-mobile/firefox-android/blob/5aac35da96f041c8f78358fd44e8af87cd35a46d/android-components/components/feature/downloads/src/main/java/mozilla/components/feature/downloads/AbstractFetchDownloadService.kt#L190-L199)



I would rather prefer that we update `handleRemovePrivateDownloadIntent` which was not cancelling  the `DownloadJob`and reuse our existing functionally  for [ACTION_CANCEL](https://github.com/mozilla-mobile/firefox-android/blob/5aac35da96f041c8f78358fd44e8af87cd35a46d/android-components/components/feature/downloads/src/main/java/mozilla/components/feature/downloads/AbstractFetchDownloadService.kt#L190-L199)
 than duplicating it.

Sharing a possible approach: 

```diff
diff --git a/android-components/components/feature/downloads/src/main/java/mozilla/components/feature/downloads/AbstractFetchDownloadService.kt b/android-components/components/feature/downloads/src/main/java/mozilla/components/feature/downloads/AbstractFetchDownloadService.kt
index b67a8c7bde..4f14e03e61 100644
--- a/android-components/components/feature/downloads/src/main/java/mozilla/components/feature/downloads/AbstractFetchDownloadService.kt
+++ b/android-components/components/feature/downloads/src/main/java/mozilla/components/feature/downloads/AbstractFetchDownloadService.kt
@@ -187,16 +187,7 @@ abstract class AbstractFetchDownloadService : Service() {
                     }
 
                     ACTION_CANCEL -> {
-                        removeNotification(context, currentDownloadJobState)
-                        currentDownloadJobState.lastNotificationUpdate = System.currentTimeMillis()
-                        setDownloadJobStatus(currentDownloadJobState, CANCELLED)
-                        currentDownloadJobState.job?.cancel()
-
-                        currentDownloadJobState.job = CoroutineScope(IO).launch {
-                            deleteDownloadingFile(currentDownloadJobState.state)
-                            currentDownloadJobState.downloadDeleted = true
-                        }
-
+                        cancelDownloadJob(context, currentDownloadJobState)
                         removeDownloadJob(currentDownloadJobState)
                         emitNotificationCancelFact()
                         logger.debug("ACTION_CANCEL for ${currentDownloadJobState.state.id}")
@@ -242,6 +233,21 @@ abstract class AbstractFetchDownloadService : Service() {
         }
     }
 
+    private fun cancelDownloadJob(
+        context: Context,
+        currentDownloadJobState: DownloadJobState,
+    ) {
+        removeNotification(context, currentDownloadJobState)
+        currentDownloadJobState.lastNotificationUpdate = System.currentTimeMillis()
+        setDownloadJobStatus(currentDownloadJobState, CANCELLED)
+        currentDownloadJobState.job?.cancel()
+
+        currentDownloadJobState.job = CoroutineScope(IO).launch {
+            deleteDownloadingFile(currentDownloadJobState.state)
+            currentDownloadJobState.downloadDeleted = true
+        }
+    }
+
     override fun onBind(intent: Intent?): IBinder? = null
 
     override fun onCreate() {
@@ -279,6 +285,7 @@ abstract class AbstractFetchDownloadService : Service() {
     internal fun handleRemovePrivateDownloadIntent(download: DownloadState) {
         if (download.private) {
             downloadJobs[download.id]?.let {
+                cancelDownloadJob(context, it)
                 removeDownloadJob(it)
             }
             store.dispatch(DownloadAction.RemoveDownloadAction(download.id))
```

Comment (iorgamgabriel) [reply android-components/components/feature/downloads/src/main/java/mozilla/components/feature/downloads/AbstractFetchDownloadService.kt]: Done

**Review by Amejia481 — APPROVED** [review]
🥇

**Pull request merged**